### ISSUE TITLE
Accessibility Level 1 Improvements

### ISSUE THREAD

**Issue by trebuan202** [gchq/stroom#5397]
- [ ] - [5398 - Non-text content accessibility improvements](https://github.com/gchq/stroom/issues/5398) 
 - [ ] - [5399 - Info and relationships accessibility improvements](https://github.com/gchq/stroom/issues/5399)
 - [ ] - [5400 - keyboard accessibility improvements](https://github.com/gchq/stroom/issues/5400)
 - [ ] - [5401 - No keyboard trap accessibility improvements](https://github.com/gchq/stroom/issues/5401)
 - [ ] - [5402 - Character key shortcuts accessibility improvements](https://github.com/gchq/stroom/issues/5402)
 - [ ] - [5403 - Bypass blocks accessibility improvements](https://github.com/gchq/stroom/issues/5403)
 - [ ] - [5404 - Page titled accessibility improvements](https://github.com/gchq/stroom/issues/5404)
 - [ ] - [5405 - Focus order accessibility improvements](https://github.com/gchq/stroom/issues/5405)
 - [ ] - [5406 - Link purpose accessibility improvements](https://github.com/gchq/stroom/issues/5406)
 - [ ] - [5407 - Label in name accessibility improvements](https://github.com/gchq/stroom/issues/5407)
 - [ ] - [5408 - Language of page accessibility improvements](https://github.com/gchq/stroom/issues/5408)
 - [ ] - [5409 - Error identification accessibility improvements](https://github.com/gchq/stroom/issues/5409)
 - [ ] - [5410 - Labels or instructions accessibility improvements](https://github.com/gchq/stroom/issues/5410)
 - [ ] - [5411 - Name, role, value accessibility improvements](https://github.com/gchq/stroom/issues/5411)

1.1.1 Non-Text Content (Partially Supports)

Within a Stroom Pipeline, each pipeline element contains an icon to highlight what type of element they represent ( e.g. filter, XSL etc ), however, there is no textual alternative for this icon. The screen reader therefore is unable to give any textual representation of this icon.

1.3.1 Info and Relationships (Partially Supports)
The main screen of Stroom contains a menu in the left-hand panel. It is not recommended to use tables because structural parts of the table ( e.g. row, column) are included in the screen reader dialogue. Alternatively, use a list which would also contain extra detail to highlight the purpose of the menu.

2.1.1 Keyboard (Partially Supports)
Keyboard navigation should flow from top-to-bottom, left-to-right and when you come across a selectable item you should be able to open that item using <enter> or <space bar>.
With keyboard navigation, the hamburger menu ( 3 dots ) which is located in the top-right of the main screen web page, forms part of the top menu of the central panel. When you keyboard navigate across the top menu of the central panel, it does not flow directly to the hamburger menu. Keyboard navigation from the top menu of the central panel moves to the content of the central panel first as opposed to moving to the hamburger menu first. 

With keyboard navigation and selectable items. In the left-hand panel of the main screen, in the case where it includes selectable items ( e.g. Template Pipelines ) you should be able to select or expand this item using a keyboard control like <Enter> or <Space>. You are unable to open such items using a keyboard control.

With keyboard navigation and popup dialog boxes, when these are selected then navigation should subsequently flow within that dialog box. When selecting the “quick syntax filter” help option within the quick filter field in the left-panel, it displays the help text within a popup dialog box, but this dialog box does not get keyboard focus. You are unable therefore to read any of the text within that dialog box.  

2.1.1 No Keyboard Trap (Partially Supports)
When you navigate to the hamburger menu ( 3 dots ) which is located in the top-right of the main screen, you are unable to navigate forward. I would expect that navigation either take you back to the left panel items or it would take you to the browsers address bar. Either way, there should be no keyboard trap here. 

2.1.4 Character Key Shortcuts (Does not support )
In the case where character-only shortcuts are used, these can be inappropriate and frustrating for speech input users. Within Stroom, it provides single character shortcuts such as W-A-S-D and H-J-K-L for movement. There does not appear to be a facility to turn these off or remap them to some other key sequence.

2.4.1 Bypass Blocks (Partially supports)
Whilst Stroom provides a wide range of shortcuts to enable you to get direct access to different screens and to enable you to perform actions, there were some areas where you were unable to bypass repeated content. As the selectable menu options were located in the left panel you would use these to indicate your selection. In the case where there were existing items ( e.g. pipeline, XSL etc), any selected item would appear in the central panel. Using the keyboard, in order to move to the central panel to inspect the selected item one had to navigate through repeated content.  

2.4.2 Page Titled (Partially supports)
 This requires that each web page has a unique title. The main title in the StroomUI is a single generic “Stroom” title. The title could be changed in order to reflect what functional area the user is currently working on. For example, when you select a functional area like “user groups” or “pipeline” or “Feed” then the you could change the title to include that functional area e.g. “Stroom – pipeline steps”.

2.4.3 Focus Order (Partially supports)

When items are selected using the keyboard ( e.g. security, user groups ), focus should be given to that viewable item. Currently, in Stroom, when you select an item ( e.g. security, user groups ) focus remains on the main menu. Also, when inside the viewable item ( e.g. security, user groups ) focus order using the keyboard keys does not follow a sequential ( left-to-right, top-down ) approach.

2.4.4 Link Purpose (In Context) (Partially supports)
There are some links whose link description could be improved. For example, in the left-hand menu panel there, are short-cut links/options named “New” and “Delete”, the alternative text for these could have some context.

2.5.3 Label in Name (Partially Supports)
There are some areas in Stroom where the visible name is not the same as the accessibility name which is picked up by the screen reader. For example, in the left hand panel which can contain expandable items ( e.g Template Pipelines), when this item is visually displayed as expanded, the accessibility name for these expanded items are of the form “row 1 group 1 column 1”, “row 2 group 1 column1” etc. In the example of the template pipelines, these label names are visually shown as “batch search”, “context data” etc.
 
3.1.1 Language of Page ( Does not support)

The default language used within the Stroom is English, however, there is no programmatic indication of this. To explicitly indicate to a screen reader and other accessibility tools, what the main language is, the “lang” attribute needs to be set within the  HTML element.

3.3.1 Error Identification ( Partially supports)
In functional areas like “find in content” or “main menu->navigation -> find”, more guidance could be given here as to what is valid input.

Also, when creating a new Feed, if you don’t put a value in the “Name” field, an “Alert” pop-up dialog box is displayed visually with the message “You must provide a name for the feed”. However, to a screen reader this error message is not spoken and focus is moved to the “Close” button.  

Also, in the “Create user” option “security->users->create user”, when using just keyboard controls, it allows you to create a user when all the fields are left blank. Visually, it appears that the field labelled “Unique user identity” is mandatory but no error message is given. Also, using just keyboard controls when choosing the options “Okay” and “Cancel” appear to have the same outcome.

3.3.2 Labels or instructions ( Partially supports )

Generally, Stroom provides good accessibility field names/labels across its input forms. However, more attention could be given to highlight whether input fields are mandatory or optional. For example, whenever, you select the option “New” ( to create a feed, pipeline etc ) it doesn’t indicate whether a “Name” field is mandatory. This is a problem across most input forms.  

4.1.2 Name, Role, Value ( Partially supports)

It is expected that elements within a Web page each have a name, role and value which are accessible by a screen reader. 

 There are some areas within Stroom where name, role and value could be improved. 

With role for example, whilst standard widget roles such as button, checkbox etc are present, some custom items ( e.g. those representing Stroom items like Feed, Pipeline ) don’t have a role. From an accessibility perspective, these could be assigned a document role such as “feed”, “pipeline” etc.

Also, within areas of Stroom which use a checkbox ( e.g. users list, the enabled column; user groups, application and document permissions), it wasn’t clear from an accessibility perspective what the value of these checkboxes were. In general, name, role and value could be added to such items.
